src/Network/Wai/Middleware/Logging.hs

```diff
@@ -0,0 +1,114 @@
+-- | Log requests (more accurately, responses) as they happen
+--
+-- In JSON format, logged messages look like:
+--
+-- @
+-- {
+--   ...
+--   message: {
+--     text: "GET /foo/bar => 200 OK",
+--     meta: {
+--       method: "GET",
+--       path: "/foo/bar",
+--       query: "?baz=bat&quix=quo",
+--       status: {
+--         code: 200,
+--         message: "OK"
+--       },
+--       durationMs: 1322.2,
+--       requestHeaders: {
+--         Authorization: "***",
+--         Accept: "text/html",
+--         Cookie: "***"
+--       },
+--       responseHeaders: {
+--         Set-Cookie: "***",
+--         Expires: "never"
+--       }
+--     }
+--   }
+-- }
+-- @
+--
+module Network.Wai.Middleware.Logging
+  ( requestLogger
+  ) where
+
+import Prelude
+
+import Control.Arrow ((***))
+import Control.Monad.IO.Unlift (withRunInIO)
+import Data.Aeson
+import qualified Data.Aeson.Compat as Key
+import qualified Data.Aeson.Compat as KeyMap
+import qualified Data.CaseInsensitive as CI
+import Data.Text (pack)
+import Data.Text.Encoding (decodeUtf8)
+import Logging
+import Network.HTTP.Types.Header (Header, HeaderName)
+import Network.HTTP.Types.Status (Status(..))
+import Network.Wai
+  ( Middleware
+  , Request
+  , Response
+  , rawPathInfo
+  , rawQueryString
+  , requestHeaders
+  , requestMethod
+  , responseHeaders
+  , responseStatus
+  )
+import qualified System.Clock as Clock
+
+requestLogger :: HasLogger env => env -> Middleware
+requestLogger env app req respond =
+  runLoggerLoggingT env $ withRunInIO $ \runInIO -> do
+    begin <- getTime
+    app req $ \resp -> do
+      recvd <- respond resp
+      duration <- toMillis . subtract begin <$> getTime
+      recvd <$ runInIO (logResponse duration req resp)
+ where
+  getTime = Clock.getTime Clock.Monotonic
+
+  toMillis x = fromIntegral @_ @Double (Clock.toNanoSecs x) / 10000
+
+logResponse :: MonadLogger m => Double -> Request -> Response -> m ()
+logResponse duration req resp
+  | statusCode status >= 500 = logError $ message :# details
+  | statusCode status == 404 = logDebug $ message :# details
+  | statusCode status >= 400 = logWarn $ message :# details
+  | otherwise = logDebug $ message :# details
```

Comment:
I wonder if these mappings should be configurable someday...

Reply:
Yeah, I can definitely see a settings type and `requestLogger = requestLoggerWith someDefaults` in the future.